include/rcutils/shared_library.h

```diff
@@ -0,0 +1,79 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCUTILS__SHARED_LIBRARY_H_
+#define RCUTILS__SHARED_LIBRARY_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <string.h>
+
+#ifndef _WIN32
+#include <dlfcn.h>
+#else
+#include <windows.h>
+#endif  // _WIN32
+
+#include "rcutils/macros.h"
+#include "rcutils/visibility_control.h"
+
+/// The structure
+typedef struct rcutils_shared_library_t
+{
+  /// The pointer to the shared library.
+  #ifndef _WIN32
+  void * lib_pointer;
+  #else
+  HINSTANCE lib_pointer;
+  #endif
+  /// The path of the shared_library
+  char * library_path;
+} rcutils_shared_library_t;
+
+/// Return shared library pointer.
+/**
+ * \param[in] library_path path to the library
+ * \return void* shared library pointer.
+ *                nullptr if library doesn't exist
+ */
+RCUTILS_PUBLIC
+rcutils_shared_library_t *
+rcutils_get_shared_library(const char * library_path);
```

Comment:
I think it would be better to have this signature:

```c
rcutils_ret_t
rcutils_get_shared_library(rcutils_shared_library_t * shared_library, const char * library_path);
```

For two reasons:

- this puts the effort of allocating the `rcutils_shared_library_t` onto the user (they can choose heap or stack)
- it allows you to return more precise error codes if that makes sense

Reply:
If you do this, I recommend a function like `rcutils_shared_library_t rcutils_get_zero_initialized_shared_library()` which the user may use to ensure the struct is initialized (even if we add a field in the future). For example:

https://github.com/ros2/rcutils/blob/af3e14543ffa7e8d4d9d38f9c242548f76accd89/include/rcutils/types/string_map.h#L55-L56